tracing-tracy/src/lib.rs

```diff
@@ -144,6 +145,26 @@ impl Default for TracyLayer {
     }
 }
 
+static MAX_CACHE_SIZE: AtomicUsize = AtomicUsize::new(usize::MAX);
+
+/// Specify the maximum number of bytes used in thread local caches.
+///
+/// A value of zero disables the cache, while a value of [`usize::MAX`] denotes an unlimited
+/// cache size.
+///
+/// Note: the upper bound on the cache size is respected on a best effort basis only. We make
+/// no guarantees on the maximum memory used by tracing-tracy. Notably, changes to this value
+/// are eventually consistent, i.e. caches are not flushed upon an update.
+///
+/// Defaults to [`usize::MAX`].
```

Comment:
Hm, `usize::MAX` is a poor default, I feel. As I said before, it is pretty straightforward to use `tracing` in a way that would lead people to indefinitely keep allocating strings in one of the threads and then release them in another. I would also argue that people by default won’t go invoking this function to set a finite limit proactively when integrating `tracing-tracy` into their codebase. When that happens, it can be pretty difficult to actually troubleshoot and figure out where the memory is getting leaked to when it starts happening (most likely in production.)

On the other hand "tracing is not going fast enough" is much easier to investigate (you can easily narrow it down with some directed experiments.)

Reply:
I also am left wondering if there’s much value in trying to track the size/capacity of the strings contained within the cache (rather than just a number of elements.)

One reason why I’m saying this is because if the code spawns N threads, `tracing-tracy` will ultimately end up using N*bytes space, and N is largely unbounded.

And then tracking bytes has a potential issue in that if the cache ends up with a single very large string inside it, it can end up pushing out many small strings out, which would end up likely hurting the throughput much more.

Reply:
> Hm, usize::MAX is a poor default

Fair enough, how about 8K?

> allocating strings in one of the threads and then release them in another

Hmmm, hadn't thought of this. Wouldn't it be pretty unusual to never re-use allocations on the receiver thread though? Here's how I'm thinking spans can move across threads:
- Async: each await point leads to potential thread switching.
- Spawn a background task and move the span into the closure. This would be the case where you don't get to re-use allocations, but it seems weird to me that you'd explicitly start a span on one thread and move into another since your timing between the task being scheduled and starting will be pretty arbitrary.
- Anything else I'm missing?

If we want to focus on the background task case (which I'm not sure we should unless it's more common than I think), then that's where using your ArrayQueue seems like a great idea.

> I also am left wondering if there’s much value in trying to track the size/capacity of the strings contained within the cache (rather than just a number of elements.)

Yeah, I spent some time thinking about this too and didn't really get anywhere. I'm not sure I agree that N is unbounded in practice. I can see how thread counts could increase burstly, but when the thread dies its TLS gets de-allocated so we're fine.

My primary concern with simply using the number of strings is that if you have a single span that has HUGE fields for whatever reason, you'll be stuck with this massive allocation forever whereas the max bytes approach will probably kill it.

> And then tracking bytes has a potential issue in that if the cache ends up with a single very large string inside it, it can end up pushing out many small strings out

Yup, this is really annoying and bothered me too. My best idea to fix this is to sort the vec by string size every time we trim it. But then I wasn't sure which strings to get rid of... nuking small strings means one very large string can dominate while nuking large strings means you lose end up realloc-ing more. If you think this is a good idea, I think the best approach is probably nuking the biggest strings since that will bias towards having a lot of medium sized strings (I think).

Reply:
Yeah, in async scenarios you are almost definitely not going to have the same allocation but it is also quite guaranteed that there is going to be a reasonably good random distribution of which thread either end gets executed by.

The 2nd case with "spawning a background task" is the concerning one. I actually do something quite like it at $DAYJOB (though moving into threads of rayon pool) in order to establish parent-child relationship between spans of the parent thread and the work that gets executed on the rayon pool. Otherwise the span gets "detached" and the stuff that happens within rayon is not as actionable.

Perhaps something that’s unique to rayon threads is that work that happens within those is largely "scoped". I.e. work that happens within is guaranteed to start and finish within the boundaries of the parent span, only that there can be multiple parallel ones. But it is also a perfect example of where the movement of span ownership is very directional.

Here's a [basic example](https://github.com/near/nearcore/blob/4942ba29f295f7e7b6e452c84000bb48aee79262/tools/mock-node/src/setup.rs#L199)

Reply:
I'm a little confused, in the example provided isn't the parent span allocated and de-allocated on the same thread since it waits for all the tasks to be finished? Seems like it'd work perfectly in this scenario.

Reply:
Ah right, my bad. This is not `move` closure. I did distinctly remember them being passed around by value and such, but the code must have changed since. And indeed scoped threads can use references if they wish to do so.

So the concerning part is really just moving a span into a different (possibly non-scoped) thread. How likely is that kinda use? Not sure. Is it possible? Yes, definitely. Are there use-cases that doing so serves? Context.

Reply:
> moving a span into a different (possibly non-scoped) thread. How likely is that kinda use?

I feel like pretty low. Or at least this code doesn't make that case much worse (it'll take up extra space allocating the vec but that's it). If we wanted to avoid any downside from this code, we'd need to add a max thread local bytes used API and then setting it to 0 makes these changes behavior the same as before. We totally could, but I feel like it would be a pretty niche use case. Do you want me to add that API, or should we wait for someone who needs it?

Reply:
My only requirement is that there isn’t an infinite memory leak (or unsoundness, or...) for that kind of code. I’m okay with things not being perfect, say, perf-wise for likely rare corner-cases like these.

Reply:
Ah, cool cool! Then let me go with the 8K default limit and biggest string eviction policy.

Reply:
Done.